File: cider/src/main/java/com/hartwig/hmftools/cider/AsyncBamReader.java
package com.hartwig.hmftools.cider;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.NoSuchElementException;
import java.util.Queue;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.function.BiConsumer;

import com.hartwig.hmftools.common.genome.region.GenomeRegion;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import htsjdk.samtools.SAMRecord;
import htsjdk.samtools.SAMRecordIterator;
import htsjdk.samtools.SamReader;
import htsjdk.samtools.SamReaderFactory;

public class AsyncBamReader
{
    private static final Logger logger = LogManager.getLogger(AsyncBamReader.class);

    // use inheritance to save some memory, it is a bit dirty
    // but Amber is running out of memory
    static class Task<E extends GenomeRegion>
    {
        E genomeRegion;
    }

    static class BamReaderThread<E extends GenomeRegion> extends Thread
    {
        final Queue<Task<E>> mTaskQ;
        final SamReader mSamReader;
        final int mMinMappingQuality;
        BiConsumer<E, SAMRecord> mConsumer;

        BamReaderThread(final String bamFile, final SamReaderFactory samReaderFactory, final Queue<Task<E>> inTaskQ,
                int minMappingQuality, BiConsumer<E, SAMRecord> consumer)
        {
            mTaskQ = inTaskQ;
            mSamReader = samReaderFactory.open(new File(bamFile));
            mMinMappingQuality = minMappingQuality;
            mConsumer = consumer;
        }

        @Override
        public void run()
        {
            logger.debug("bam reader thread start");

            while (true)
            {
                Task<E> task;
                try
                {
                    task = mTaskQ.remove();
                }
                catch (NoSuchElementException e)
                {
                    // finished processing
                    break;
                }

                try (final SAMRecordIterator iterator = task.genomeRegion == null ?
                        mSamReader.queryUnmapped() :
                        mSamReader.queryOverlapping(task.genomeRegion.chromosome(), task.genomeRegion.start(), task.genomeRegion.end()))
                {
                    while (iterator.hasNext())
                    {
                        final SAMRecord record = iterator.next();

                        if (!passesFilters(record))
                        {
                            continue;
                        }

                        //if (task.genomeRegion.start() >= record.getAlignmentStart() && task.genomeRegion.end() <= alignmentEnd)
                        if (task.genomeRegion.start() <= record.getAlignmentEnd() && task.genomeRegion.end() >= record.getAlignmentStart())
                        {
                            mConsumer.accept(task.genomeRegion, record);
                        }
                    }
                }
            }

            try {
                mSamReader.close();
            }
            catch (IOException e)
            {
                logger.error("IO exception in SamReader::close: {}", e.getMessage());
            }

            logger.debug("bam reader thread finish");
        }

        private boolean passesFilters(final SAMRecord record)
        {
            if(record.getMappingQuality() < mMinMappingQuality || record.getReadUnmappedFlag())
                return false;

            return !record.getDuplicateReadFlag();
        }
    }

    public static <E extends GenomeRegion> void processBam(final String bamFile, final SamReaderFactory samReaderFactory,
            final Collection<E> regions, BiConsumer<E, SAMRecord> asyncRecordHandler, int threadCount, int minMappingQuality)
            throws InterruptedException
    {
        logger.debug("Processing {} potential sites in bam {}", regions.size(), bamFile);

        final Queue<Task<E>> taskQ = new ConcurrentLinkedQueue<>();
        
        // create genome regions from the loci
        populateTaskQueue(regions, taskQ);
        BamTaskCompletion taskCompletion = new BamTaskCompletion(taskQ.size());

        // we create the consumer and producer
        var bamReaders = new ArrayList<BamReaderThread<E>>();

        for (int i = 0; i < Math.max(threadCount, 1); ++i)
        {
            var t = new BamReaderThread<>(bamFile, samReaderFactory, taskQ, minMappingQuality, asyncRecordHandler);
            t.setName(String.format("worker-%d", i));
            t.start();
            bamReaders.add(t);
        }

        logger.info("{} bam reader threads started", bamReaders.size());

        for (BamReaderThread<E> t : bamReaders)
        {
            while (t.isAlive())
            {
                // check status every 30 seconds
                t.join(30_000);

                // check status
                taskCompletion.progress(taskQ.size());
            }
        }

        logger.info("{} bam reader threads finished", bamReaders.size());
    }

    /**
     * Group the genome regions
     */
    public static <E extends GenomeRegion> void populateTaskQueue(final Collection<E> genomeRegions, final Queue<Task<E>> taskQ)
    {
        for (E genomeRegion : genomeRegions)
        {
            Task<E> task = new Task<>();
            task.genomeRegion = genomeRegion;
            taskQ.add(task);
        }
        // another one for unmapped reads
        taskQ.add(new Task<>());
    }
}

File: cider/src/main/java/com/hartwig/hmftools/cider/AsyncBamReader.kt
package com.hartwig.hmftools.cider

import com.hartwig.hmftools.common.genome.region.GenomeRegion
import htsjdk.samtools.SAMRecord
import htsjdk.samtools.SAMRecordIterator
import htsjdk.samtools.SamReader
import htsjdk.samtools.SamReaderFactory
import org.apache.logging.log4j.LogManager
import java.io.File
import java.io.IOException
import java.util.*
import java.util.concurrent.BlockingQueue
import java.util.concurrent.LinkedBlockingDeque

object AsyncBamReader
{
    private const val MAX_BAM_RECORD_Q_SIZE = 100000
    private val logger = LogManager.getLogger(AsyncBamReader::class.java)

    @JvmStatic
    @Throws(InterruptedException::class)
    fun processBam(
        bamFile: String,
        samReaderFactory: SamReaderFactory,
        genomeRegions: Collection<GenomeRegion>,
        asyncRecordHandler: (SAMRecord) -> Unit,
        threadCount: Int
    )
    {
        logger.debug("Processing {} potential sites in bam {}", genomeRegions.size, bamFile)

        // create genome regions from the loci
        // val taskCompletion = BamTaskCompletion(taskQ.size)

        val bamRecordQ: BlockingQueue<Optional<SAMRecord>> = LinkedBlockingDeque()

        // create the bam record consumers
        val recordConsumers = ArrayList<BamRecordConsumerThread>()

        for (i in 0 until Math.max(threadCount, 1))
        {
            val t = BamRecordConsumerThread(bamRecordQ, asyncRecordHandler)
            t.name = String.format("worker-%d", i)
            t.start()
            recordConsumers.add(t)
        }
        logger.info("{} bam record consumer threads started", recordConsumers.size)

        val bamReader = BamReader(bamFile, samReaderFactory, genomeRegions, bamRecordQ)
        bamReader.run()
        bamRecordQ.put(Optional.empty()) // signals consumer to finish

        for (t in recordConsumers)
        {
            t.join()
        }

        logger.info("{} bam reader threads finished", recordConsumers.size)
    }

    internal class BamReader(
        bamFile: String,
        samReaderFactory: SamReaderFactory,
        private val genomeRegionList: Collection<GenomeRegion>,
        private val outputRecordQ: Queue<Optional<SAMRecord>>)
    {
        private val mSamReader: SamReader = samReaderFactory.open(File(bamFile))

        fun run()
        {
            logger.debug("bam reader start")

            for (genomeRegion in genomeRegionList)
            {
                logger.debug("querying genome region: {}", genomeRegion)
                mSamReader.queryOverlapping(genomeRegion.chromosome(), genomeRegion.start(), genomeRegion.end())
                    .use({ iterator -> processRecords(iterator) })
            }

            // we do not process unmapped reads
            // mSamReader.queryUnmapped().use({ iterator -> processRecords(iterator) })

            try
            {
                mSamReader.close()
            }
            catch (e: IOException)
            {
                logger.error("IO exception in SamReader::close: {}", e.message)
            }
            logger.debug("bam reader finish")
        }

        private fun processRecords(iterator: SAMRecordIterator)
        {
            while (iterator.hasNext())
            {
                val record = iterator.next()

                if (record.duplicateReadFlag)
                {
                    // drop duplicate reads
                    continue
                }

                // we don't want to check the alignment region, reason is that we intentionally
                // want to process unmapped read where mate pairs are mapped to an interesting region
                // the downstream processing will take care of it.
                outputRecordQ.add(Optional.of(record))

                // important: we want to take a break if we are going to flood the output queue
                checkAndWaitForConsumer()
            }
        }

        private fun checkAndWaitForConsumer()
        {
            if (outputRecordQ.size >= MAX_BAM_RECORD_Q_SIZE)
            {
                logger.info("bam record Q size: {}, max size reached, pausing bam reader", outputRecordQ.size)

                // drain at least 80% of the queue before going back to work
                while (outputRecordQ.size * 5 >= MAX_BAM_RECORD_Q_SIZE)
                {
                    try
                    {
                        Thread.sleep(1000)
                    }
                    catch (ex: InterruptedException)
                    {
                        Thread.currentThread().interrupt()
                    }
                }
                logger.info("finished sleeping, bam record Q size: {}", outputRecordQ.size)
            }
        }
    }

    internal class BamRecordConsumerThread(
        private val bamRecordQ: BlockingQueue<Optional<SAMRecord>>,
        private val samRecordHandler: (SAMRecord) -> Unit
    ) : Thread()
    {
        override fun run()
        {
            logger.debug("bam record consumer thread start")
            while (true)
            {
                val record: Optional<SAMRecord> = try
                {
                    bamRecordQ.take()
                }
                catch (e: InterruptedException)
                {
                    break
                }
                if (record.isEmpty)
                {
                    // if record is empty, it signals consumer to stop
                    // we want to put it back so other consumer threads
                    // will also know to stop
                    bamRecordQ.put(record)
                    break
                }
                samRecordHandler(record.get())
            }
            logger.debug("bam record consumer thread finish")
        }
    }
}